File: Assets/PixelPlanets/StellarBreaker/Scripts/UI/MainPresenter.cs
using System;
using StellarBreaker.Config;
using StellarBreaker.Gameplay;

namespace StellarBreaker.UI
{
    /// <summary>Immutable snapshot the main-screen View renders (no Unity types → testable).</summary>
    public struct MainViewModel
    {
        public string stageLabel;
        public bool   isBoss;
        public bool   bossActive;
        public int    bossSecondsLeft;

        public string hpText;
        public float  hpFraction;        // 0..1

        public string stardustText;
        public string tapDamageText;
        public string fleetDpsText;

        public string tapUpgradeCostText;
        public bool   canUpgradeTap;

        public bool   hasShip;
        public string shipButtonText;
        public bool   canBuyShip;

        public SkillVm[] skills;
        public bool      canPrestige;
        public string    prestigeText;

        public string      relicsText;
        public ArtifactVm[] artifacts;
    }

    /// <summary>One skill button's display state.</summary>
    public struct SkillVm
    {
        public string label;
        public bool   unlocked;
        public bool   ready;
        public bool   active;
        public int    secondsLeft;   // active time left, or cooldown left
    }

    /// <summary>One artifact button's display state.</summary>
    public struct ArtifactVm
    {
        public string label;
        public string levelText;     // e.g. "Lv 3  +15%"
        public string costText;
        public bool   canBuy;
    }

    /// <summary>
    /// Maps a GameSession to a MainViewModel. The View calls Build() each frame and
    /// pushes values into Text/Image components. All numbers via BigNumber.ToShortString().
    /// </summary>
    public class MainPresenter
    {
        readonly GameSession _s;

        public MainPresenter(GameSession session) => _s = session;

        public MainViewModel Build()
        {
            var p  = _s.Enemy.Current;
            var vm = new MainViewModel
            {
                stardustText       = _s.Wallet.Stardust.ToShortString(),
                tapDamageText      = _s.TapUpgrade.CurrentDamage.ToShortString(),
                fleetDpsText       = _s.Ships.FleetDps().ToShortString(),
                tapUpgradeCostText = _s.TapUpgrade.NextCost.ToShortString(),
                canUpgradeTap      = _s.Wallet.CanAfford(_s.TapUpgrade.NextCost),
                bossActive         = _s.Stage.BossActive,
                bossSecondsLeft    = (int)Math.Ceiling(_s.Stage.BossTimeLeft),
            };

            if (p != null)
            {
                vm.stageLabel = "Stage " + p.Stage;
                vm.isBoss     = p.IsBoss;
                vm.hpText     = p.CurrentHP.ToShortString() + " / " + p.MaxHP.ToShortString();
                vm.hpFraction = (float)p.HpFraction01();
            }
            else
            {
                vm.stageLabel = "—";
                vm.hpText     = "";
                vm.hpFraction = 0f;
            }

            var ships = _s.Ships;
            if (ships.Count > 0)
            {
                vm.hasShip = true;
                string state = ships.IsOwned(0) ? "Lv " + ships.LevelOf(0) : "Buy";
                vm.shipButtonText = ships.Def(0).shipName + "  " + state + "  (" + ships.NextCost(0).ToShortString() + ")";
                vm.canBuyShip = _s.Wallet.CanAfford(ships.NextCost(0));
            }
            else
            {
                vm.hasShip = false;
                vm.shipButtonText = "—";
                vm.canBuyShip = false;
            }

            var slots = _s.SkillSlots;
            var sk = new SkillVm[slots.Count];
            for (int i = 0; i < slots.Count; i++)
            {
                var t = slots[i];
                bool active = _s.Skills.IsActive(t);
                double secs = active ? _s.Skills.ActiveTimeLeft(t) : _s.Skills.Cooldown(t);
                sk[i] = new SkillVm
                {
                    label       = _s.Skills.Name(t),
                    unlocked    = _s.Skills.IsUnlocked(t),
                    ready       = _s.Skills.CanActivate(t),
                    active      = active,
                    secondsLeft = (int)Math.Ceiling(secs),
                };
            }
            vm.skills       = sk;
            vm.canPrestige  = _s.CanPrestige();
            vm.prestigeText = "PRESTIGE  +" + _s.PreviewRelics().ToShortString();

            vm.relicsText = _s.Prestige.Relics.Stardust.ToShortString();
            var arts = _s.Artifacts;
            var av = new ArtifactVm[arts.Count];
            for (int i = 0; i < arts.Count; i++)
            {
                ArtifactDefinition d = arts.Def(i);
                int lvl = arts.LevelOf(i);
                int pct = (int)Math.Round(lvl * d.bonusPerLevel * 100.0);
                av[i] = new ArtifactVm
                {
                    label     = d.displayName,
                    levelText = "Lv " + lvl + "  +" + pct + "%",
                    costText  = arts.NextCost(i).ToShortString(),
                    canBuy    = _s.Prestige.Relics.CanAfford(arts.NextCost(i)),
                };
            }
            vm.artifacts = av;
            return vm;
        }
    }
}

File: Telegram/TelegramApp/src/game/ui/MainPresenter.ts
// Ported from Assets/PixelPlanets/StellarBreaker/Scripts/UI/MainPresenter.cs
import { BigNumber } from '../core/BigNumber'
import type { GameSession } from '../gameplay/GameSession'
import type { SkillType } from '../config/SkillDefinition'
import { artifactBonusAt } from '../config/ArtifactDefinition'

/** One skill button's display state. */
export interface SkillVm {
  type: SkillType
  label: string
  description: string
  unlocked: boolean
  ready: boolean
  active: boolean
  /** active time left, or cooldown left */
  secondsLeft: number
  /** full duration of whichever timer secondsLeft is counting down (active duration or cooldown) - for cooldown-wipe rendering */
  totalSeconds: number
}

/** One artifact button's display state. */
export interface ArtifactVm {
  label: string
  /** e.g. "Lv 3  +15%" */
  levelText: string
  costText: string
  canBuy: boolean
}

/** Immutable snapshot the main screens render (no game-object refs -> easy to test). */
export interface MainViewModel {
  stageLabel: string
  isBoss: boolean
  bossActive: boolean
  bossSecondsLeft: number

  hpText: string
  /** 0..1 */
  hpFraction: number

  stardustText: string
  tapDamageText: string
  tapLevel: number
  fleetDpsText: string

  tapUpgradeCostText: string
  canUpgradeTap: boolean

  hasShip: boolean
  shipButtonText: string
  canBuyShip: boolean

  skills: SkillVm[]
  canPrestige: boolean
  prestigeText: string

  relicsText: string
  artifacts: ArtifactVm[]

  // Progressive disclosure (FTUE): hide systems until they become relevant.
  /** after first gold/kill */
  showUpgradeTap: boolean
  /** first ship owned OR close to affordable */
  showFleet: boolean
  /** relics exist or an artifact is owned */
  showArtifacts: boolean
  /** near/at the unlock stage, or relics exist */
  showPrestige: boolean
}

/** Maps a GameSession to a MainViewModel, mirroring MainPresenter.Build() from Unity. */
export function buildMainViewModel(s: GameSession): MainViewModel {
  const p = s.enemy.current

  const skills: SkillVm[] = s.skillSlots.map((t) => {
    const active = s.skills.isActive(t)
    const secs = active ? s.skills.activeTimeLeft(t) : s.skills.cooldown(t)
    const total = active ? s.skills.fullDuration(t) : s.skills.fullCooldown(t)
    return {
      type: t,
      label: s.skills.name(t),
      description: s.skills.description(t),
      unlocked: s.skills.isUnlocked(t),
      ready: s.skills.canActivate(t),
      active,
      secondsLeft: Math.ceil(secs),
      totalSeconds: total,
    }
  })

  const arts = s.artifacts
  const artifacts: ArtifactVm[] = Array.from({ length: arts.count }, (_, i) => {
    const def = arts.def(i)
    const lvl = arts.levelOf(i)
    const pct = Math.round(artifactBonusAt(def, lvl) * 100)
    return {
      label: def.displayName,
      levelText: `Lv ${lvl}  +${pct}%`,
      costText: arts.nextCost(i).toShortString(),
      canBuy: s.prestige.relics.canAfford(arts.nextCost(i)),
    }
  })

  const ships = s.ships
  const hasShip = ships.count > 0
  const shipButtonText = hasShip
    ? `${ships.def(0).shipName}  ${ships.isOwned(0) ? `Lv ${ships.levelOf(0)}` : 'Buy'}  (${ships.nextCost(0).toShortString()})`
    : '—'

  const progressed = s.tapUpgrade.level > 1 || s.stage.currentStage > 1 || s.stage.highestStage > 1 || s.wallet.balance.gt(BigNumber.Zero)

  const anyShipOwned = ships.count > 0 && ships.fleetDps().gt(BigNumber.Zero)
  const shipClose = ships.count > 0 && s.wallet.balance.mul(new BigNumber(2)).gte(ships.nextCost(0)) // >=50% of first cost
  const showFleet = anyShipOwned || shipClose

  const anyArtifactOwned = Array.from({ length: arts.count }, (_, i) => arts.levelOf(i) > 0).some(Boolean)
  const hasRelics = s.prestige.relics.balance.gt(BigNumber.Zero)
  const showArtifacts = hasRelics || anyArtifactOwned

  const canPrestige = s.canPrestige()
  const showPrestige = canPrestige || hasRelics || s.stage.highestStage >= s.prestigeUnlockStage - 2

  return {
    stageLabel: p ? `Sector ${p.stage}` : '—',
    isBoss: p ? p.isBoss : false,
    bossActive: s.stage.bossActive,
    bossSecondsLeft: Math.ceil(s.stage.bossTimeLeft),

    hpText: p ? `${p.currentHp.toShortString()} / ${p.maxHp.toShortString()}` : '',
    hpFraction: p ? p.hpFraction01() : 0,

    stardustText: s.wallet.balance.toShortString(),
    tapDamageText: s.tapUpgrade.currentDamage.toShortString(),
    tapLevel: s.tapUpgrade.level,
    fleetDpsText: s.ships.fleetDps().toShortString(),

    tapUpgradeCostText: s.tapUpgrade.nextCost.toShortString(),
    canUpgradeTap: s.wallet.canAfford(s.tapUpgrade.nextCost),

    hasShip,
    shipButtonText,
    canBuyShip: hasShip && s.wallet.canAfford(ships.nextCost(0)),

    skills,
    canPrestige,
    prestigeText: `ASCEND  +${s.previewRelics().toShortString()}`,

    relicsText: s.prestige.relics.balance.toShortString(),
    artifacts,

    showUpgradeTap: progressed,
    showFleet,
    showArtifacts,
    showPrestige,
  }
}
